File: 2020/Day11/Python/SeatingSystem-Part2.py
import sys
import collections
import itertools
import numpy as np


def countNeigbours(world, x, y):
    if world[y, x] == ".":
        return "."
    seen = 0
    for dy in range(-1, 2):
        for dx in range(-1, 2):
            if dx == dy == 0:
                continue
            multiply = 1
            while (x + dx*multiply not in [-1, len(world[0])]) and (y + dy*multiply not in [-1, len(world)]):
                if world[y + dy*multiply, x + dx*multiply] == "#":
                    seen += 1
                    break
                elif world[y + dy*multiply, x + dx*multiply] == "L":
                    break
                multiply += 1
    if world[y, x] == "L" and seen == 0:
        return "#"
    elif world[y, x] == "#" and seen >= 5:
        return "L"
    else:
        return world[y, x]
    
world = []

for line in open(f"{sys.path[0]}/../input.txt").read().splitlines():
    world.append(list(line))

world = np.array(world)
previous = -1
while previous != (world == "#").sum():
    previous = (world == "#").sum()
    print(previous)
    newWorld = np.array(world, copy=True)
    for y in range(len(world)):
        for x in range(len(world[y])):
            newWorld[y, x] = countNeigbours(world, x, y)
    world = np.array(newWorld)

print(previous)

File: 2020/Day11/Haskell/SeatingSystem-Part2.hs
import System.IO
import Control.Monad

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | null xs = show x
    | otherwise = show x ++ "\n" ++ printArray xs

countNeighbours:: [[Char]] -> Int -> Int -> Char
countNeighbours world x y 
    | seat == '.' = '.'
    | seat == 'L' && adjacent == 0 = '#'
    | seat == '#' && adjacent >= 5 = 'L'
    | otherwise = seat
    where     
        dirs = [(i, j) | i <- [-1..1], j <- [-1..1], (i,j) /= (0,0)]
        adjacent = sum (map (\dir -> seeTaken [(x+dx*mult, y+dy*mult) | let (dx, dy) = dir, mult <- [1..]]) dirs)
        seat = world!!y!!x
        seeTaken [] = 0
        seeTaken ((x,y):xs)
            | y < 0 || y >= length world = 0
            | x < 0 || x >= length (head world) = 0
            | char == '.' = seeTaken xs
            | char == '#' = 1
            | otherwise = 0
            where 
                char = world!!y!!x

solve:: [[Char]] -> Int -> [Int]
solve world previous
    | previous == current = [previous]
    | otherwise = current : solve newWorld current
    where 
        current = sum $ map (length . filter (=='#')) world
        newWorld = [[countNeighbours world x y | x <- [0..(length (head world)-1)]] | y <- [0..length world -1]]

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let world = lines contents
    print (solve world 2)
    hClose handle